pkg/kates/accumulator_test.go

```diff
@@ -82,3 +93,72 @@ func TestBootstrapNotifyEvenOnEmptyWatch(t *testing.T) {
 	// that satisfy the selector, the ConfigMaps field should be empty.
 	assert.Equal(t, 0, len(snap.ConfigMaps))
 }
+
+// Make sure we coalesce raw changes before sending an update when a batch of resources
+// are created/modified in quick succession.
+func TestBatchNoNotifyBeforeSync(t *testing.T) {
+	ctx, cli := testClient(t, nil)
+	// Set a long interval to make sure all changes are batched before sending.
+	cli.MaxAccumulatorInterval(10 * time.Second)
+	acc, err := cli.Watch(ctx, Query{Name: "ConfigMaps", Kind: "ConfigMap", LabelSelector: "test=test-batch"})
+	require.NoError(t, err)
+
+	snap := &Snap{}
+
+	// Listen for changes from the Accumulator. Here it will listen for only 2 updates
+	// The first update should be the one sent during bootstrap. No resources should have changed
+	// in this update. The second update should contain resource changes.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		numUpdates := 0
+		for {
+			<-acc.Changed()
+			updated, err := acc.Update(ctx, snap)
+			require.NoError(t, err)
+			if updated {
+				numUpdates++
+			}
+			if numUpdates == 2 {
+				break
+			}
+		}
+	}()
+
+	// Use a separate client to create resources to avoid any potential uses of the cache
+	_, cli2 := testClient(t, nil)
+
+	// Create a set of 10 Configmaps after the Accumulator is watching to simulate getting
+	// a bunch of resources at once mid-watch.
+	var cms [10]*ConfigMap
+	for i := 0; i < 10; i++ {
+		cm := &ConfigMap{
+			TypeMeta: TypeMeta{
+				Kind: "ConfigMap",
+			},
+			ObjectMeta: ObjectMeta{
+				Name: fmt.Sprintf("test-batch-%d", i),
+				Labels: map[string]string{
+					"test": "test-batch",
+				},
+			},
+		}
+		err := cli2.Upsert(ctx, cm, cm, &cm)
+		require.NoError(t, err)
+		cms[i] = cm
+	}
+	wg.Wait()
+
+	// After receiving 2 updates from the Accumulator, we should have 10 ConfigMaps
+	// in our Snapshot due to the Accumulator coalescing changes before sending an update.
+	assert.Equal(t, 10, len(snap.ConfigMaps))
```

Comment:
Should there be an assertion after bootstrap and before we get the "mid" stream batched items? Or is this test saying that a single bootstrap will not occur because we will wait the interval and at the end of that interval you will on only get 1 "snapshot notification" (aka reconfigure) which will include all 10 items?

Reply:
that is an oversight on my part - I should be also verifying that the first change event representing bootstrap is empty